api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java

```diff
@@ -1096,4 +1105,84 @@ public DbUser confirmPublications() {
         dbUser,
         Agent.asUser(dbUser));
   }
+
+  /** Send an Access Renewal Expiration or Warning email to the user, if appropriate */
+  @Override
+  public void maybeSendAccessExpirationEmail(DbUser user) {
+    // TODO combine with CT expiration logic when available to send AT MOST ONE email
+
+    final Optional<Timestamp> rtExpiration = getRegisteredTierExpirationForEmails(user);
+    rtExpiration.ifPresent(expiration -> maybeSendRegisteredTierExpirationEmail(user, expiration));
+  }
+
+  /**
+   * Return the user's registered tier access expiration time, for the purpose of sending an access
+   * renewal reminder or expiration email.
+   *
+   * <p>First: ignore any bypassed modules. These are in compliance and do not need to be renewed.
+   *
+   * <p>Next: do all un-bypassed modules have expiration times? If yes, return the min (earliest).
+   * If no, either the feature flag is not set or the user does not have access for reasons other
+   * than access renewal compliance. In either negative case, we should not send an email.
+   *
+   * <p>Note that this method may return EMPTY for both valid and invalid users, so this method
+   * SHOULD NOT BE USED FOR ACCESS DECISIONS.
+   */
+  private Optional<Timestamp> getRegisteredTierExpirationForEmails(DbUser user) {
+    // Collection<Optional<T>> is usually a code smell.
+    // Here we do need to know if any are EMPTY, for the next step.
+    final Set<Optional<Timestamp>> expirations =
+        getRenewableAccessModules(user).values().stream()
+            .filter(times -> !times.isBypassed())
+            .map(ModuleTimes::getExpiration)
+            .collect(Collectors.toSet());
+
+    // if any un-bypassed modules are incomplete, we know:
+    // * this user does not currently have access
+    // * this user has never previously had access
+    // therefore: the user is neither "expired" nor "expiring" and we should not send an email
+
+    if (!expirations.stream().allMatch(Optional::isPresent)) {
+      return Optional.empty();
+    } else {
+      return expirations.stream()
+          .map(Optional::get)
+          // note: min() returns EMPTY if the stream is empty at this point,
+          // which is also an indicator that we should not send an email
+          .min(Timestamp::compareTo);
+    }
+  }
+
+  // TODO config values
+  final Set<Long> warningThresholds = ImmutableSet.of(1L, 3L, 7L, 15L, 30L);
```

Comment:
LMK if you want me to config-ify this, or if this is fine for now

Reply:
IMO this falls under 6703.  I'll add it to the AC there but I'll merge this now.